core/src/main/java/org/apache/accumulo/core/data/KeyBuilder.java

```diff
@@ -0,0 +1,300 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.accumulo.core.data;
+
+
+import org.apache.hadoop.io.Text;
+
+/**
+ * A builder used to build {@link Key}s by defining their components.
+ *
+ * The rules are:
+ * <ul>
+ *   <li>All components of the {@link Key} are optional except the row</li>
+ *   <li>Components not explicitly set default to empty byte array except the timestamp which
+ *   defaults to <code>Long.MAX_VALUE</code></li>
+ *   <li>The column qualifier can only be set if the column family has been set first</li>
+ *   <li>The column visibility can only be set if at least the column family has been set first</li>
+ * </ul>
+ *
+ * The builder supports three types of components: <code>byte[]</code>, <code>Text</code> and <code>CharSequence</code>.
+ *
+ * The builder is mutable and not thread safe.
+ *
+ * @see org.apache.accumulo.core.data.Key
+ * @since 1.8
+ */
+public class KeyBuilder {
+
+  public interface Build {
```

Comment:
Not 100% sure about this, but I think the `@since` tag from the enclosing class will not show up in the javadoc for this class.  So these inner interfaces may needs `@since` tags.